### ISSUE TITLE
Only events from NameResolver-produced address should trigger resolution refresh

### ISSUE THREAD

**Issue by zhangkun83** [grpc/grpc-java#1599]
Currently a failure on any `TransportSet` will trigger name resolution. However, the addresses of a `TransportSet` may also originate from an external load-balancing service, e.g., in `GrpclbLoadBalancer`. A failure on such `TransportSet` shouldn't trigger name resolution.

**Comment by zhangkun83** [comment]
This is resolved in the [LBv2](https://docs.google.com/document/d/1exE0_QNwbHJesD-Opl2D6qX1wujADkUm5OpFykks3C8/edit) ManagedChannelImpl. It needs a unit test though.

**Comment by ejona86** [comment]
@zhangkun83, I don't think this is resolved by LBv2. I think we decided we needed to expose `refresh()` explicitly to the LB and have LB propagate as appropriate. Is that correct?

**Comment by ejona86** [comment]
No, I don't know the cause of #3588, but it does make sense it is necessary. For gRPCLB we need to occasionally do refreshes to get updated addresses for the OOB Channel. In that case we don't actually care about the backend disconnects... except for things like the TXT of service config...

**Comment by zhangkun83** [comment]
#3588 is for b/67375167. I was asking the background of #1599. If you don't know either, I guess we can close it as WAI.

**Comment by ejona86** [comment]
Oh, the background for this one is that disconnects on grpclb backends will currently trigger NR refreshes, but those aren't expected to be useful. We can just rely on disconnects for connections to the grpclb server and in the fallback mode.

**Comment by zhangkun83** [comment]
No, the first comment on this one is saying disconnects for connections to the grpclb server **should not** trigger resolution refresh, which is the opposite of what you said.

**Comment by ejona86** [comment]
It said (emphasis mine):
> _addresses_ of a TransportSet may also _originate_ from an external load-balancing service

So that's talking about backend addresses returned by the grpclb service **should not** trigger refresh. That's the same case I'm talking about.

Disconnects for connections to the grpclb server **should** trigger refresh, as done in #3588.

**Comment by zhangkun83** [comment]
You are right... So we will stick to the plan about exposing `refresh()` to `LoadBalancer`s.

**Comment by zhangkun83** [comment]
We can make ManagedChannelImpl to attach the source NameResolver as an attribute to the EAGs from the NameResolver before passing them to LoadBalancer. Assuming LoadBalancers don't discard the EAG attributes before passing them to `createSubchannel()` and `createOobChannel()`. ManagedChannelImpl can then refresh the source NameResolver whenever an InternalSubchannel is disconnected, whether it's an Subchannel or an OobChannel. By doing this, LoadBalancers don't need to refresh name resolution at all.

To support the service-routing case where a parent balancer creates child NameResolver/LoadBalancer pairs, we'd need to pack the NameResolver/LoadBalancer pair into a class, say "RequestDirector", that would absorb `getNameResolver()` and `NameResolverListenerImpl` (which would include the NameResolver-attaching logic) from `ManagedChannelImpl`.

**Comment by ejona86** [comment]
There would be some strangeness for pick-first, since the subchannel will have multiple EAGs. But that seems resolvable.

**Comment by ejona86** [comment]
Fixed by #9429
